waku/waku_archive/retention_policy/retention_policy_size.nim

```diff
@@ -0,0 +1,98 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/times,
+  stew/results,
+  chronicles,
+  chronos,
+  os
+import
+  ../driver,
+  ../retention_policy
+
+logScope:
+  topics = "waku archive retention_policy"
+
+# default size is 30 Gb
+const DefaultRetentionSize*: float = 30_720
+
+# to remove 20% of the outdated data from database
+const DeleteLimit = 0.80
+
+type
+  # SizeRetentionPolicy implements auto delete as follows:
+  # - sizeLimit is the size in megabytes (Mbs) the database can grow upto
+  # to reduce the size of the databases, remove the rows/number-of-messages
+  # DeleteLimit is the total number of messages to delete beyond this limit
+  # when the database size crosses the sizeLimit, then only a fraction of messages are kept,
+  # rest of the outdated message are deleted using deleteOldestMessagesNotWithinLimit(),
+  # upon deletion process the fragmented space is retrieve back using Vacuum process. 
+  SizeRetentionPolicy* = ref object of RetentionPolicy
+      sizeLimit: float
+
+proc init*(T: type SizeRetentionPolicy, size=DefaultRetentionSize): T =
+  SizeRetentionPolicy(
+    sizeLimit: size
+  )
+
+method execute*(p: SizeRetentionPolicy,
+                driver: ArchiveDriver):
+                Future[RetentionPolicyResult[void]] {.async.} =
+  ## when db size overshoots the database limit, shread 20% of outdated messages 
+  
+  # to get the size of the database, pageCount and PageSize is required
+  # get page count in "messages" database
+  var pageCount = (await driver.getPagesCount()).valueOr:
+    return err("failed to get Pages count: " & $error)
+
+  # get page size of database
+  var pageSizeRes = await driver.getPagesSize()
+  var pageSize: int64 = int64(pageSizeRes.valueOr(0) div 1024)
+
+  if pageSize == 0:
+    return err("failed to get Page size: " & pageSizeRes.error)
+
+  # database size in megabytes (Mb)
+  var totalSizeOfDB: float = float(pageSize * pageCount)/1024.0
+
+  # check if current databse size crosses the db size limit
+  if totalSizeOfDB < p.sizeLimit:
+    return ok()
+
+  # keep deleting until the current db size falls within size limit 
```

Comment:
I see in this function a pattern that is usually solved with a `do while`. A "do while" can remove aroud 25 lines of code, since `totalSizeOfDB`, `pageSize` and `pageCount` are a bit boilerplate.

Since nim doesnt have "do while" I would suggest:

```nim
while true:
    ...(get page get count)
    var totalSizeOfDB: float = float(pageSize * pageCount)/1024.0
    if totalSizeOfDB > p.sizeLimit:
      break
    let res = await driver.deleteOldestMessagesNotWithinLimit(limit=pageDeleteWindow)
    if res.isErr():
        return err("deleting oldest messages failed: " & res.error)
```

So with something like this you just need 1 single call to `totalSizeOfDB`, `pageSize` and `pageCount`